Add circular visual timer and task tasking functionality

rebeccamccabe opened the pull request:
<img width="1366" height="1087" alt="image" src="https://github.com/user-attachments/assets/87b84e68-905a-4347-8529-14cf63ce0d3f" />

- In pomodoro mode, adds a colored visual timer (circular arc that decreases in angle as time elapses, changing from green to yellow to red based on how much time is left)
- Changes the timer end notification in pomodoro mode to be modal, requiring user acknowledgement before being able to access the rest of vscode, to ensure timer is not ignored
- Adds task tracking feature, where upon timer start, users are prompted for what task they are working on, and the task is displayed in the sidebar to help remember what you are doing

rebeccamccabe committed:
baseline working visual timer

rebeccamccabe committed:
fix z layer issue where visual timer overlaid countdown

rebeccamccabe committed:
relative width so visual timer expands to window width

rebeccamccabe committed:
change color green to yellow to red

rebeccamccabe committed:
auto-open to pomo tab instead of stopwatch

rebeccamccabe committed:
change finish notification to be modal to block progress until acknowledged

rebeccamccabe committed:
add task tracking feature to prompt and display a task

rebeccamccabe committed:
remove Math.floor to make visual timer update smooth instead of ticking

Dj-Artimus reviewed (changes requested):
Hi @rebeccamccabe ,

Thanks so much for your pull request! I genuinely appreciate you taking the time to contribute these fantastic new features to Time Coder, especially the visual timer, modal notification, and task tracking. These are all significant enhancements that will greatly improve the user experience!

I've had a chance to review the changes and test them out thoroughly. Overall, it's a very strong contribution. I've identified a few areas where we can refine things further to ensure consistency, optimal user experience across different themes, and full feature integration. Please see the detailed points below:

## 1. Visual Timer Display Issues

### A. Shape Distortion of the Circular Progress Bar

I've noticed that as the timer progresses, especially around the 60% mark, the circular shape of the progress indicator seems to distort, appearing more rectangular or jagged. It should ideally maintain a smooth, circular arc throughout its entire progression.

Here are a few examples of the distortion:

* **Initial State (Good):**
<img width="457" height="986" alt="image" src="https://github.com/user-attachments/assets/be01b821-f1b6-4a83-91bb-b0f9387bf9ba" />

* **Distortion Example 1 (Around 20-30 seconds remaining):**
<img width="464" height="975" alt="image" src="https://github.com/user-attachments/assets/ee2eb07c-e9d9-4122-8866-ad6562dc33d3" />

* **Distortion Example 2 (Getting close to end):**
<img width="439" height="953" alt="image" src="https://github.com/user-attachments/assets/0393fce4-37d5-44fc-8fe9-5268724c0d8c" />

Could you please investigate why this might be happening and ensure it always renders as a perfect segment of a circle?

### B. Visibility on Dark Themes (Timer Text vs. Progress Circle)

The intensity of the progress circle's color (particularly the yellow and orange) sometimes clashes with the white timer text on dark VS Code themes. This makes the numerical timer difficult to read due to low contrast. We need to ensure excellent contrast for readability.

* **Low Contrast Example:**
<img width="425" height="918" alt="image" src="https://github.com/user-attachments/assets/7c83b132-90ba-4c58-a0c6-0229b64728c7" />

Could we perhaps reduce the opacity slightly or adjust the color saturation for dark themes to ensure the white text remains clearly visible? This might involve using VS Code's theme-aware CSS variables if possible, or detecting the current theme and adjusting styles accordingly.

### 2. Task Title Visibility on Dark Themes

The task tracking feature is excellent! However, similar to the main timer, the visibility of the 'No task' text (and presumably the actual task title when entered) in the sidebar has a visibility issue on dark themes. It blends in too much with the background.

* **'No task' on Dark Theme:** 
<img width="422" height="950" alt="image" src="https://github.com/user-attachments/assets/fbc23bea-8a74-4d50-87a6-4bb25afdf822" />

Please adjust the text color for dark themes to ensure it stands out clearly against the dark background.

### 3. Timer Title Consistency

I appreciate the creative flair with titles like 'Pom⏰odoro', but I'd like to maintain the original naming conventions for the timers ('Pomodoro', 'Stopwatch', 'Session Tracker'). My original intention for 'Pomodoro' was 'Pom⏰doro', with the clock icon representing the 'o'.

Could you please revert these titles to their original spelling?

### 4. Webview Dashboard Integration

The webview dashboard is a key part of the extension for users who prefer a larger, dedicated view. It would be excellent if we could integrate the new colored progress circle visual timer into the webview dashboard as well. Currently, it's only appearing in the sidebar.

* **Current Webview State (No Progress Circle):**
<img width="1868" height="976" alt="image" src="https://github.com/user-attachments/assets/91c1aa75-1b84-41b6-94b7-d4efb555fc8e" />

Please extend the implementation to include the animated progress circle in the webview view too, maintaining consistency with the sidebar.

---

Thank you again for your hard work on this pull request! I'm confident that addressing these points will make Time Coder even more polished and user-friendly. Please let me know if you have any questions or need clarification on any of these points.

Looking forward to the next iteration!

Best regards,
@Dj-Artimus

rebeccamccabe commented:
Thanks for the feedback! I will incorporate the requested changes in the
next few days.

-Becca


On Tue, Jul 22, 2025 at 5:00 PM Dj-Artimus ***@***.***> wrote:

> ***@***.**** requested changes on this pull request.
>
> Hi @rebeccamccabe <https://github.com/rebeccamccabe> ,
>
> Thanks so much for your pull request! I genuinely appreciate you taking
> the time to contribute these fantastic new features to Time Coder,
> especially the visual timer, modal notification, and task tracking. These
> are all significant enhancements that will greatly improve the user
> experience!
>
> I've had a chance to review the changes and test them out thoroughly.
> Overall, it's a very strong contribution. I've identified a few areas where
> we can refine things further to ensure consistency, optimal user experience
> across different themes, and full feature integration. Please see the
> detailed points below:
> 1. Visual Timer Display Issues A. Shape Distortion of the Circular
> Progress Bar
>
> I've noticed that as the timer progresses, especially around the 60% mark,
> the circular shape of the progress indicator seems to distort, appearing
> more rectangular or jagged. It should ideally maintain a smooth, circular
> arc throughout its entire progression.
>
> Here are a few examples of the distortion:
>
>    - *Initial State (Good):*
>    ![Initial State - Green]
>
> image.png (view on web)
> <https://github.com/user-attachments/assets/be01b821-f1b6-4a83-91bb-b0f9387bf9ba>
>
>    - *Distortion Example 1 (Around 20-30 seconds remaining):*
>    ![Distortion 1 - Yellow]
>
> image.png (view on web)
> <https://github.com/user-attachments/assets/ee2eb07c-e9d9-4122-8866-ad6562dc33d3>
>
>    - *Distortion Example 2 (Getting close to end):*
>    ![Distortion 2 - Red]
>
> image.png (view on web)
> <https://github.com/user-attachments/assets/0393fce4-37d5-44fc-8fe9-5268724c0d8c>
>
> Could you please investigate why this might be happening and ensure it
> always renders as a perfect segment of a circle?
> B. Visibility on Dark Themes (Timer Text vs. Progress Circle)
>
> The intensity of the progress circle's color (particularly the yellow and
> orange) sometimes clashes with the white timer text on dark VS Code themes.
> This makes the numerical timer difficult to read due to low contrast. We
> need to ensure excellent contrast for readability.
>
>    - *Low Contrast Example:*
>    ![Low Contrast Example]
>
> image.png (view on web)
> <https://github.com/user-attachments/assets/7c83b132-90ba-4c58-a0c6-0229b64728c7>
>
> Could we perhaps reduce the opacity slightly or adjust the color
> saturation for dark themes to ensure the white text remains clearly
> visible? This might involve using VS Code's theme-aware CSS variables if
> possible, or detecting the current theme and adjusting styles accordingly.
> 2. Task Title Visibility on Dark Themes
>
> The task tracking feature is excellent! However, similar to the main
> timer, the visibility of the 'No task' text (and presumably the actual task
> title when entered) in the sidebar has a visibility issue on dark themes.
> It blends in too much with the background.
>
>    - *'No task' on Dark Theme:*
>
> image.png (view on web)
> <https://github.com/user-attachments/assets/fbc23bea-8a74-4d50-87a6-4bb25afdf822>
>
> Please adjust the text color for dark themes to ensure it stands out
> clearly against the dark background.
> 3. Timer Title Consistency
>
> I appreciate the creative flair with titles like 'Pom⏰odoro', but I'd like
> to maintain the original naming conventions for the timers ('Pomodoro',
> 'Stopwatch', 'Session Tracker'). My original intention for 'Pomodoro' was
> 'Pom⏰doro', with the clock icon representing the 'o'.
>
> Could you please revert these titles to their original spelling?
> 4. Webview Dashboard Integration
>
> The webview dashboard is a key part of the extension for users who prefer
> a larger, dedicated view. It would be excellent if we could integrate the
> new colored progress circle visual timer into the webview dashboard as
> well. Currently, it's only appearing in the sidebar.
>
>    - *Current Webview State (No Progress Circle):*
>
> image.png (view on web)
> <https://github.com/user-attachments/assets/91c1aa75-1b84-41b6-94b7-d4efb555fc8e>
>
> Please extend the implementation to include the animated progress circle
> in the webview view too, maintaining consistency with the sidebar.
> ------------------------------
>
> Thank you again for your hard work on this pull request! I'm confident
> that addressing these points will make Time Coder even more polished and
> user-friendly. Please let me know if you have any questions or need
> clarification on any of these points.
>
> Looking forward to the next iteration!
>
> Best regards,
> @Dj-Artimus <https://github.com/Dj-Artimus>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Dj-Artimus/time-coder/pull/1#pullrequestreview-3044863227>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHUJPSWMBPDO27PDUCOK5U33J2Q47AVCNFSM6AAAAACCA6UYJCVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTANBUHA3DGMRSG4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
